### PR TITLE
Re-implement project() with a (PETSc)solver

### PR DESCRIPTION
**Pull request by nate-sime**
Simple update for `dolfin/fem/projection` to work with the PETSc solvers.

`project()` now takes a `solver` optional argument, which is used to solve the underlying linear system.

Also added `dolfin::la::PETScVector::sum, min, max, abs` to the python layer.

### CODE REVIEW

File: python/src/la.cpp [hunk 1]
```diff
@@ -674,6 +674,10 @@ void la(py::module& m)
       .def("update_ghost_values", &dolfin::la::PETScVector::update_ghost_values)
       .def("size", (std::size_t(dolfin::la::PETScVector::*)() const)
                        & dolfin::la::PETScVector::size)
+      .def("sum", &dolfin::la::PETScVector::sum)
```
Comment (garth-wells) [inline python/src/la.cpp]: Do we want these wrappers, or just use petsc4py? @chrisrichardson ?

Comment (nate-sime) [reply python/src/la.cpp]: I'm happy for these to be removed, depending on @chrisrichardson's opinion also.

File: python/test/unit/function/test_projection.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+                         dtype=numpy.double)
+    f_comp = f_comp[d2v]
+    ownership_range = V.dofmap().ownership_range()
+    owned_offset = ownership_range[1] - ownership_range[0]
+    f_comp = f_comp[:owned_offset]
+
+    assert numpy.allclose(f_comp - f_arr, 0.0)
+
+
+@pytest.mark.parametrize('solver_factory', solver_factories)
```
[the remark this answers is not in the document]
Comment (nate-sime) [reply python/test/unit/function/test_projection.py]: I like to be sure the interface works. The extra cost is small. If you want them gone they can be removed.

[the remark this answers is not in the document]
Comment (nate-sime) [reply python/test/unit/function/test_projection.py]: And now there's only one type of solver, the parameterisation can be eliminated.

**Pull request closed without merging**